src/sc2api/sc2_client.cc

```diff
@@ -2035,73 +2039,60 @@ void ControlImp::IssueUnitDestroyedEvents() {
-            client_.OnUnitEnterVision(&unit);
-        }
-        else if (unit.alliance == Unit::Alliance::Self) {
-            client_.OnUnitCreated(&unit);
-        }
-    });
-}
 
-void ControlImp::IssueIdleEvent(const Unit* unit, const std::vector<Tag>& commands) {
-    if (!unit || !unit->orders.empty() || unit->build_progress < 1.0f) {
-        return;
+    for (auto unit : observation_imp_->unit_pool_.GetNewUnits())
+    {
+        if (unit->alliance == Unit::Alliance::Self)
+            client_.OnUnitCreated(unit);
+        else if (unit->alliance == Unit::Alliance::Neutral && unit->display_type == Unit::DisplayType::Visible)
+            client_.OnNeutralUnitCreated(unit);
     }
-    // Lookup unit from previous map.
-    auto found = observation_imp_->units_previous_map_.find(unit->tag);
 
-    // If it's not in the previous map it's a new unit with new orders so trigger the OnIdle event.
-    if (found == observation_imp_->units_previous_map_.end()) {
-        client_.OnUnitIdle(unit);
-        return;
+    for (auto unit : observation_imp_->unit_pool_.GetUnitsEnteringVision())
+    {
+        if (unit->alliance == Unit::Alliance::Enemy && unit->display_type == Unit::DisplayType::Visible)
+            client_.OnUnitEnterVision(unit);
     }
+}
 
-    // Otherwise get that unit from the previous list and verify it's state changed to idle.
-    const Unit& unit_previous = found->second;
-
-    if (!unit_previous.orders.empty()) {
-        client_.OnUnitIdle(unit);
-        return;
-    }
+void ControlImp::IssueUnitDamagedEvents() {
+    for (auto const *u : observation_imp_->unit_pool_.GetDamagedUnits())
+        client_.OnUnitDamaged(u);
+}
 
-    // If the unit had less than 1.0 build progress in the last stop this is the first time it's active.
-    if (unit_previous.build_progress < 1.0f) {
-        client_.OnUnitIdle(unit);
-        return;
-    }
+void ControlImp::IssueIdleEvents(const std::vector<Tag>& commands) {
 
-    // Iterate the issued commands, if a unit exists in that list but does not currently have orders
-    // the order must have failed. Reissue the OnUnitIdle event in that case.
+    auto& unit_pool = observation_imp_->unit_pool_;
+    // identify idled units where commands were issued last step, but units have no orders now (maybe failed, maybe executed instantly)
     for (auto t : commands) {
-        if (t == unit->tag) {
-            client_.OnUnitIdle(unit);
-            break;
-        }
+        const auto* unit = unit_pool.GetExistingUnit(t);
+        if (unit && unit->orders.empty())
+            unit_pool.AddUnitIdled(unit);
     }
-}
 
-void ControlImp::IssueBuildingCompletedEvent(const Unit* unit) {
-    // If the units build progress is complete but it previously wasn't call construction complete
-    if (!unit || unit->build_progress < 1.0f) {
-        return;
-    }
+    // add newly created units (if they are completed)
+    for (auto const* u : unit_pool.GetNewUnits())
```

Comment:
Style: please add `{}` to the cycle body. Here and in other places.

Reply:
done